### PR TITLE
fix: (Core) add enhancement for dynamic Carousel items and example file rename

### PR DESCRIPTION
**Pull request by DeepakSap14**
#### Please provide a link to the associated issue.
#3881
#3956 
#### Please provide a brief summary of this pull request.
1. When carousel used inside search result page. carousel needs to be updated when carousel items changes dynamically.
2. Changed Documentation file names and corresponding component name as per standard formats.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 1]
```diff
@@ -249,6 +251,7 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
             this.navigation = false;
         }
 
+        this.slides.changes.subscribe(() => this._onSlideUpdates());
```
Comment (KevinOkamoto) [inline libs/core/src/lib/carousel/carousel.component.ts:254]: You need to close this subscription `onDestroy` of this component. Add `.pipe(takeUntil(this._onDestroy$)).subscribe(...)` as you have done elsewhere in this file. Also I noticed other subscriptions in this file where you need to do this - go through and make sure you're closing any subscriptions you have opened.

**Review by KevinOkamoto — CHANGES_REQUESTED** [review]

**Comment by mikerodonnell89** [comment]
Go to the second page and click "hide card two" - I think you should expect to see card one in that first slot but it is empty.

Screen Shot 2020-11-30 at 11 02 28 AM

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
See above comment

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 2]
```diff
@@ -391,6 +394,19 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
         }
     }
 
+    private _onSlideUpdates(): void {
```
Comment (JKMarkowski) [inline libs/core/src/lib/carousel/carousel.component.ts:397]: Can you add `/** @hidden */` comment there?

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
There are some bugs with carousel + hide/show items
carousellll

**Comment by JKMarkowski** [comment]
Button goes crazy there on press
carousellllll

**Review by Betrozov — APPROVED** [review]

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 1 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts]: this is fixed now. using pipe(takeUntil(this._onDestroy$)) now.

**Comment by DeepakSap14** [comment]
> Go to the second page and click "hide card two" - I think you should expect to see card one in that first slot but it is empty.
> 
> Screen Shot 2020-11-30 at 11 02 28 AM

I was using host binding and it has bug.
https://github.com/angular/angular/issues/22560

It is fixed now.

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 2 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts:417]: added.

**Comment by DeepakSap14** [comment]
> Button goes crazy there on press
> carousellllll

This issue is coming from fundamental-styles. i have raised issue to address it.
This is only happening for chrome browser.
https://github.com/SAP/fundamental-styles/issues/1927

**Comment by InnaAtanasova** [comment]
> > Button goes crazy there on press
> > carousellllll
> 
> This issue is coming from fundamental-styles. i have raised issue to address it.
> This is only happening for chrome browser.
> [SAP/fundamental-styles#1927](https://github.com/SAP/fundamental-styles/issues/1927)

I opened a PR for this in styles: https://github.com/SAP/fundamental-styles/pull/1928

**Review by KevinOkamoto — APPROVED** [review]

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 3]
```diff
@@ -238,8 +242,12 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
...
             this.rightButtonDisabled = true;
         }
-        this._carouselService.activeChange.subscribe((event) => this._onSlideSwipe(event));
-        this._carouselService.dragStateChange.subscribe((event) => this._onSlideDrag(event));
+        this._carouselService.activeChange
+            .pipe(takeUntil(this._onDestroy$))
+            .subscribe((event) => this._onSlideSwipe(event));
+        this._carouselService.dragStateChange
+            .pipe(takeUntil(this._onDestroy$))
+            .subscribe((event) => this._onSlideDrag(event));
```
Comment (JKMarkowski) [inline libs/core/src/lib/carousel/carousel.component.ts:250]: Could you move it to separated method?

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 4]
```diff
@@ -249,6 +257,8 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
             this.navigation = false;
         }
```
Comment (JKMarkowski) [inline libs/core/src/lib/carousel/carousel.component.ts:258]: Could you move this statement to line 242? I see you check for slides in both places.

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
@DeepakSap14  I found 2 minor things in code. Examples work well for me. Great job!

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 3 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts]: moved to separate method.

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 4 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts:248]: done

**Comment by DeepakSap14** [comment]
@JKMarkowski please review now.

**Review by mikerodonnell89 — APPROVED** [review]

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 5]
```diff
@@ -237,18 +241,18 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
...
-        this._slidesCopy = this.slides.toArray().slice();
 
-        // Disable swipe when there is no carousel item
-        if (this.slides.length === 0) {
+            // Disable swipe when there is no carousel item
             this.swipeEnabled = false;
             this.navigation = false;
         }
 
+        this._slidesCopy = this.slides.toArray().slice();
```
Comment (salarenko) [inline libs/core/src/lib/carousel/carousel.component.ts:250]: Why `.slice()`?

File: libs/core/src/lib/carousel/carousel.component.ts [hunk 6]
```diff
@@ -321,7 +325,17 @@ export class CarouselComponent implements OnInit, AfterContentInit, AfterViewIni
         this._adjustActiveItemPosition(SlideDirection.PREVIOUS);
         this._preventDefaultBtnFocus();
         this._carouselService.pickPrevious(this.dir);
-        this._notifySlideChange(SlideDirection.PREVIOUS);
+
+        /** Handle looped carousel, first click on prev button. */
+        if (this.loop && this._prevBtnClickedAtStart) {
+            this._carouselService.goToItem(this.slides.toArray()[Math.ceil(this.slides.length / 2) - 1], true, this.dir);
```
Comment (salarenko) [inline libs/core/src/lib/carousel/carousel.component.ts:331]: Looks how many times you are creating new slides array 😛
Get the value once and use it.

**Review by salarenko — DISMISSED** [review]

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 5 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts:252]: placed comment now.
In looped carousel, original array gets changes with shift and unshift. for carousel indicator, have to keep a copy of original array.

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 6 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts]: fixed.

**Review by JKMarkowski — APPROVED** [review]
MY comments have been addressed, LGTM

File: libs/core/src/lib/carousel/carousel.component.ts (hunk 5 above)
Comment (DeepakSap14) [reply libs/core/src/lib/carousel/carousel.component.ts]: removed un-necessary slice of array.

**Pull request merged**